robottelo/vm.py

```diff
@@ -214,37 +218,67 @@ def create(self):
         )
 
         result = ssh.command(
-                command,
-                self.provisioning_server,
-                connection_timeout=30
+            command,
+            self.provisioning_server,
+            connection_timeout=30
         )
-
         if result.return_code != 0:
             raise VirtualMachineError(
                 u'Failed to run snap-guest: {0}'.format(result.stderr))
         else:
             self._created = True
-
+            self.mac = [n.split('MAC:')[1].strip() for n in result.stdout if 'MAC:' in n][0]
         # outside of VLANs ping from hypervisor, in VLANs ping from SAT
         if self.bridge == 'br0':
             ping_from_hostname = self.provisioning_server
         else:
             ping_from_hostname = settings.server.hostname
 
         # Give some time to machine boot
-        result = ssh.command(
-            u'for i in {{1..60}}; do ping -c1 {0}.local && exit 0; sleep 1;'
-            u' done; exit 1'.format(self._target_image),
-            ping_from_hostname,
-            connection_timeout=30
-        )
-        if result.return_code != 0:
-            logger.error('Failed to obtain VM IP, reverting changes')
-            self.destroy()
-            raise VirtualMachineError(
-                'Failed to fetch virtual machine IP address information')
-        output = ''.join(result.stdout)
-        self.ip_addr = output.split('(')[1].split(')')[0]
+        for i in range(60):
+            qemu_ga_check = ssh.command(
+                u'virsh qemu-agent-command {0} '
+                '\'{{"execute":"guest-network-get-interfaces"}}\''.format(self.hostname),
+                ping_from_hostname,
+                connection_timeout=30
+            )
+            if qemu_ga_check.return_code != 0:
+                if 'guest agent is not connected' in qemu_ga_check.stderr:
+                    # this means the agent wasn't started yet (vm still booting)
+                    sleep(1)
+                else:
+                    # this means there's another error with agent, e.g. it is not configured
+                    break
+            else:
+                ifaces = json.loads(qemu_ga_check.stdout[0])
+                mgmt_if = [i for i in ifaces['return']
+                           if i['hardware-address'].lower() == self.mac.lower()]
+                if len(mgmt_if) > 0 and mgmt_if[0].get('ip-addresses'):
+                    # get only the ipv4 addresses
+                    ip4_addr = [i['ip-address'] for i in mgmt_if[0]['ip-addresses']
+                                if i['ip-address-type'] == 'ipv4']
+                    self.ip_addr = next(
+                        (i['ip-address'] for i in mgmt_if[0]['ip-addresses']
+                            if i['ip-address-type'] == 'ipv4'),
+                        None
+                    )
+                    if self.ip_addr:
+                        break
```

Comment:
You create `ip4_addr` variable which you never use.

This whole `else` can be rewritten. Instead of:
```python
            else:
                ifaces = json.loads(qemu_ga_check.stdout[0])
                mgmt_if = [i for i in ifaces['return']
                           if i['hardware-address'].lower() == self.mac.lower()]
                if len(mgmt_if) > 0 and mgmt_if[0].get('ip-addresses'):
                    # get only the ipv4 addresses
                    ip4_addr = [i['ip-address'] for i in mgmt_if[0]['ip-addresses']
                                if i['ip-address-type'] == 'ipv4']
                    self.ip_addr = next(
                        (i['ip-address'] for i in mgmt_if[0]['ip-addresses']
                            if i['ip-address-type'] == 'ipv4'),
                        None
                    )
                    if self.ip_addr:
                        break
```
do:
```python
            else:
                ifaces = json.loads(qemu_ga_check.stdout[0])
                mgmt_if = next((i for i in ifaces['return']
                           if i['hardware-address'].lower() == self.mac.lower()), {})
                try:
                    self.ip_addr = next(i['ip-address'] for i in mgmt_if['ip-addresses']
                            if i['ip-address-type'] == 'ipv4')
                    break
                except (KeyError, StopIteration):
                    continue
```

`KeyError` will be raised if `mgmt_if` doesn't have `ip-addresses` key (because qemu-guest didn't return us interface with appropriate MAC address or returned interface is not assigned any IP address yet). `StopIteration` will be raised if generator comprehension returns empty list (because IPv4 address is not configured). I believe this covers all your cases?

Reply:
nice!